File: ts/src/lock.ts
import { randomBytes } from "node:crypto";
import {
  closeSync,
  fchmodSync,
  fsyncSync,
  openSync,
  realpathSync,
  renameSync,
  statSync,
  unlinkSync,
  writeSync,
} from "node:fs";
import { basename, dirname, join } from "node:path";

import { EntryconfError } from "./errors.ts";

export const LOCK_FILE_NAME = ".entryconf.lock";

/** A lock file older than this is presumed abandoned by a crashed writer (SPEC §10.7). */
export const LOCK_STALE_AFTER_MS = 30_000;

/** How long `commit` waits for the lock when no timeout is given. */
export const DEFAULT_LOCK_TIMEOUT_MS = 5_000;

function sleep(ms: number): void {
  Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, ms);
}

function randomSuffix(): string {
  return randomBytes(6).toString("hex");
}

/**
 * SPEC §10.7 lock protocol: exclusive-create `<dir>/.entryconf.lock`, retry
 * until the timeout, break a lock older than `LOCK_STALE_AFTER_MS` by renaming
 * it away first (so at most one waiter breaks a given stale lock). Returns the
 * release function.
 */
export function acquireLock(dir: string, timeoutMs: number): () => void {
  const path = join(dir, LOCK_FILE_NAME);
  const deadline = Date.now() + timeoutMs;
  let wait = 5;
  for (;;) {
    try {
      const fd = openSync(path, "wx", 0o644);
      try {
        writeSync(
          fd,
          `{"pid": ${process.pid}, "created": ${JSON.stringify(new Date().toISOString())}}\n`,
        );
      } finally {
        closeSync(fd);
      }
      return () => {
        try {
          unlinkSync(path);
        } catch {
          // already gone
        }
      };
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code !== "EEXIST") {
        throw new EntryconfError(
          "E_LOCKED",
          `cannot create lock file ${path}: ${(err as Error).message}`,
        );
      }
    }
    let stale = false;
    try {
      stale = Date.now() - statSync(path).mtimeMs > LOCK_STALE_AFTER_MS;
    } catch {
      // vanished between attempts; retry
    }
    if (stale) {
      const aside = `${path}.${randomSuffix()}`;
      try {
        renameSync(path, aside);
        unlinkSync(aside);
      } catch {
        // another waiter broke it first
      }
      continue;
    }
    if (Date.now() >= deadline) {
      throw new EntryconfError(
        "E_LOCKED",
        `lock file ${path} is held by another writer (waited ${timeoutMs}ms)`,
      );
    }
    sleep(wait);
    if (wait < 100) wait *= 2;
  }
}

/**
 * Replace `target`'s content atomically (SPEC §10.7 step 3): a temporary file
 * beside the resolved target, flushed, given the target's permission bits, and
 * renamed over it. A symlinked target is resolved first so the link survives.
 * On any failure the temporary file is removed and the target is untouched.
 */
export function atomicWrite(target: string, data: Uint8Array): void {
  const fail = (step: string, err: unknown): EntryconfError =>
    new EntryconfError("E_WRITE", `${step} ${target}: ${(err as Error).message}`);

  let resolved: string;
  let mode: number;
  try {
    resolved = realpathSync(target);
    mode = statSync(resolved).mode & 0o7777;
  } catch (err) {
    throw fail("cannot resolve", err);
  }
  const dir = dirname(resolved);
  const tmp = join(dir, `.${basename(resolved)}.entryconf-tmp-${randomSuffix()}`);
  let fd: number | null = null;
  try {
    fd = openSync(tmp, "wx", 0o600);
  } catch (err) {
    throw fail("cannot create a temporary file beside", err);
  }
  try {
    writeSync(fd, data);
    fsyncSync(fd);
    if (process.platform !== "win32") fchmodSync(fd, mode);
    closeSync(fd);
    fd = null;
    renameSync(tmp, resolved);
  } catch (err) {
    if (fd !== null) {
      try {
        closeSync(fd);
      } catch {
        // ignore
      }
    }
    try {
      unlinkSync(tmp);
    } catch {
      // ignore
    }
    throw fail("cannot replace", err);
  }
  // Best effort: make the rename durable where the platform allows.
  try {
    const dfd = openSync(dir, "r");
    try {
      fsyncSync(dfd);
    } finally {
      closeSync(dfd);
    }
  } catch {
    // not supported here
  }
}

File: go/lock.go
package entryconf

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

const (
	lockFileName = ".entryconf.lock"

	// lockStaleAfter is how old a lock file must be before a waiter may treat
	// it as abandoned by a crashed writer (SPEC §10.7). A commit holds the lock
	// for milliseconds.
	lockStaleAfter = 30 * time.Second

	// DefaultLockTimeout is how long Commit waits for the directory lock when
	// CommitOptions.LockTimeout is zero.
	DefaultLockTimeout = 5 * time.Second
)

// acquireLock implements the SPEC §10.7 lock protocol: exclusive-create
// <dir>/.entryconf.lock, retrying until timeout, breaking a lock older than
// lockStaleAfter by renaming it away first so that at most one waiter breaks
// a given stale lock.
func acquireLock(dir string, timeout time.Duration) (release func(), err error) {
	path := filepath.Join(dir, lockFileName)
	deadline := time.Now().Add(timeout)
	wait := 5 * time.Millisecond
	for {
		f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o644)
		if err == nil {
			fmt.Fprintf(f, "{\"pid\": %d, \"created\": %q}\n", os.Getpid(), time.Now().UTC().Format(time.RFC3339))
			f.Close()
			return func() { os.Remove(path) }, nil
		}
		if !errors.Is(err, os.ErrExist) {
			return nil, wrapf(CodeLocked, err, "cannot create lock file %q", path)
		}
		if info, statErr := os.Stat(path); statErr == nil && time.Since(info.ModTime()) > lockStaleAfter {
			// Presumed abandoned: rename it to a unique name and delete that.
			// Only one waiter's rename succeeds, so the lock is broken once.
			stale := path + "." + randomSuffix()
			if os.Rename(path, stale) == nil {
				os.Remove(stale)
			}
			continue
		}
		if time.Now().After(deadline) {
			return nil, errf(CodeLocked, "lock file %q is held by another writer (waited %s)", path, timeout)
		}
		time.Sleep(wait)
		if wait < 100*time.Millisecond {
			wait *= 2
		}
	}
}

func randomSuffix() string {
	var buf [6]byte
	if _, err := rand.Read(buf[:]); err != nil {
		return fmt.Sprintf("%d", time.Now().UnixNano())
	}
	return hex.EncodeToString(buf[:])
}

// atomicWrite replaces target's content with data (SPEC §10.7 step 3): a
// temporary file in the target's directory, flushed, given the target's
// permission bits, and renamed over it. A symlinked target is resolved first
// so the link survives and the file it points at is what changes. On any
// failure the temporary file is removed and the target is untouched.
func atomicWrite(target string, data []byte) error {
	resolved, err := filepath.EvalSymlinks(target)
	if err != nil {
		return wrapf(CodeWrite, err, "cannot resolve %q", target)
	}
	info, err := os.Stat(resolved)
	if err != nil {
		return wrapf(CodeWrite, err, "cannot stat %q", resolved)
	}
	dir, base := filepath.Split(resolved)
	tmp, err := os.CreateTemp(dir, "."+base+".entryconf-tmp-*")
	if err != nil {
		return wrapf(CodeWrite, err, "cannot create a temporary file beside %q", resolved)
	}
	tmpPath := tmp.Name()
	fail := func(step string, cause error) error {
		tmp.Close()
		os.Remove(tmpPath)
		return wrapf(CodeWrite, cause, "%s %q", step, resolved)
	}
	if _, err := tmp.Write(data); err != nil {
		return fail("cannot write replacement for", err)
	}
	if err := tmp.Sync(); err != nil {
		return fail("cannot flush replacement for", err)
	}
	if err := tmp.Chmod(info.Mode().Perm()); err != nil {
		// Windows has no permission bits to preserve; elsewhere this matters.
		if !isWindows() {
			return fail("cannot set permissions on replacement for", err)
		}
	}
	if err := tmp.Close(); err != nil {
		os.Remove(tmpPath)
		return wrapf(CodeWrite, err, "cannot close replacement for %q", resolved)
	}
	if err := os.Rename(tmpPath, resolved); err != nil {
		os.Remove(tmpPath)
		return wrapf(CodeWrite, err, "cannot replace %q", resolved)
	}
	// Best effort: make the rename itself durable where the platform allows.
	if d, err := os.Open(dir); err == nil {
		_ = d.Sync()
		d.Close()
	}
	return nil
}
